internal/gossiphttp/transport.go

```diff
@@ -0,0 +1,554 @@
+	_ memberlist.NodeAwareTransport = (*Transport)(nil)
+)
+
+// http2Stream implements the streamClient interface and allows for sending and
+// receiving messages in a bidirectional streaming connection.
+type http2Stream struct {
+	t *Transport
+	r io.ReadCloser
+	w io.Writer
+}
+
+// Send sends b over an http2Stream connection.
+func (sc *http2Stream) Send(b []byte) error {
+	return writeMessage(sc.w, b)
+}
+
+// Recv reads from an http2Stream connection.
+func (sc *http2Stream) Recv() ([]byte, error) {
+	return readMessage(sc.r)
+}
+
+var _ io.WriteCloser = (*flushWriter)(nil)
+
+// flushWriter wraps an io.Writer with an http.Flusher to flush buffered data
+// to a streaming HTTP/2 connection's request body.
+type flushWriter struct {
+	w io.Writer
+	f http.Flusher
+}
+
+func (w *flushWriter) Write(data []byte) (int, error) {
+	n, err := w.w.Write(data)
+	w.f.Flush()
+	return n, err
+}
+
+func (w *flushWriter) Close() error { return nil }
+
+// NewTransport returns a new Transport. Transports must be attached to an HTTP
+// server so their endpoints are invoked. See [Handler] for more information.
+func NewTransport(opts Options) (*Transport, prometheus.Collector, error) {
+	if opts.Client == nil {
+		return nil, nil, fmt.Errorf("HTTP client must be provided")
+	}
+
+	l := opts.Log
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	t := &Transport{
+		log:     l,
+		opts:    opts,
+		metrics: newMetrics(),
+
+		inPacketQueue:  queue.New(packetBufferSize),
+		outPacketQueue: queue.New(packetBufferSize),
+
+		inPacketCh: make(chan *memberlist.Packet),
+		streamCh:   make(chan net.Conn),
+
+		exited: make(chan struct{}),
+		cancel: cancel,
+	}
+
+	t.metrics.Add(prometheus.NewGaugeFunc(
+		prometheus.GaugeOpts{
+			Name: "cluster_transport_rx_packet_queue_length",
+			Help: "Current number of unprocessed incoming packets",
+		},
+		func() float64 { return float64(t.inPacketQueue.Size()) },
+	))
+	t.metrics.Add(prometheus.NewGaugeFunc(
+		prometheus.GaugeOpts{
+			Name: "cluster_transport_tx_packet_queue_length",
+			Help: "Current number of unprocessed outgoing packets",
+		},
+		func() float64 { return float64(t.outPacketQueue.Size()) },
+	))
+
+	go t.run(ctx)
+
+	return t, t.metrics, nil
+}
+
+func (t *Transport) run(ctx context.Context) {
+	defer close(t.exited)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	defer wg.Wait()
+
+	// Close our queues before shutting down. This must be done before calling
+	// wg.Wait as it will cause the goroutines to exit.
+	defer func() { _ = t.inPacketQueue.Close() }()
+	defer func() { _ = t.outPacketQueue.Close() }()
+
+	// Process queue of incoming packets
+	go func() {
+		defer wg.Done()
+
+		for {
+			v, err := t.inPacketQueue.Dequeue(context.Background())
+			if err != nil {
+				return
+			}
+
+			pkt := v.(*memberlist.Packet)
+			t.metrics.packetRxTotal.Inc()
+			t.metrics.packetRxBytesTotal.Add(float64(len(pkt.Buf)))
+
+			t.inPacketCh <- pkt
```

Comment:
Should this not block forever if the context gets canceled?

```suggestion
			select {
			case <-ctx.Done():
			case t.inPacketCh <- pkt:
			}
```